### DISCUSSION TITLE
与鼠须管 Swift 化有关的一个难题。

### DISCUSSION THREAD

**Discussion by ShikiSuen** [rime/home#1093] [General]
我嘗試在 bridging header 裡面塞入 rime_api.h。然而，Swift 卻無法認出裡面的內容。
這個問題解決不了的話，鼠鬚管的 Swift 化只能說是不可能（除非用 Swift 徹底重寫 libRime）。
恐怕需要將目前 C++ 版 libRime 包裝成一個單獨的 Swift Package 才可以。
（我對 Swift 與 C 的 Interop 幾乎沒什麼經驗就是了。）

<img width="1920" alt="image" src="https://user-images.githubusercontent.com/3164826/177775827-fb896e64-3ea2-4f74-ba95-1660c5608c05.png">

**Comment by ShikiSuen** [comment]
靠，原來 Swift 擅自給方法重新命名了：
```
    rime_get_api()->RimeConfigOpen("squirrel", &_config)
```

**Comment by ShikiSuen** [comment]
看到 UnsafePointer 我就炸了。
以現在的我的實力水準，用 Swift 重寫 libRime 可能比較快。
看來鼠鬚管的 Swift 移植工作得交給同時精通 C 與 Swift 的人來完成。
<img width="640" alt="image" src="https://user-images.githubusercontent.com/3164826/177777666-b059d2e3-23d8-4465-a06d-1e93849ec8f1.png">

**Comment by ShikiSuen** [comment]
我把我迄今為止的工作階段放在這裡了：
https://github.com/shikisuen/squirrel/tree/Swiftification
誰感興趣的話可以在此基礎上繼續研究。

<img width="846" alt="image" src="https://user-images.githubusercontent.com/3164826/177778196-ffaf7258-e31e-435f-a24a-6427237ab957.png">

**Comment by LEOYoon-Tsaw** [comment]
讚！我有個想法，是改用Text Kit 2來繪製文字，但那是個大工程了，也許永遠不會實現

**Comment by ShikiSuen** [comment]
@LEOYoon-Tsaw TextKit2 太新，要求至少 macOS 12。

另外，Squirrel 的 Swift 化有著落了。其實只要把 RIME 的類型與函式包裝成 Objective-C 當中的可以被 Swift 沿襲使用的 NSObject 就行。

下述四個檔案應該可以派得上用場（GPLv3），不過可能需要補回最近八年內發生在 rime_api.h 這個檔案內的修改：

https://github.com/stackia/XIME/blob/master/XIME/Source/RimeTypeDef.h
https://github.com/stackia/XIME/blob/master/XIME/Source/RimeTypeDef.m
https://github.com/stackia/XIME/blob/master/XIME/Source/RimeWrapper.h
https://github.com/stackia/XIME/blob/master/XIME/Source/RimeWrapper.m
